### ISSUE TITLE
[Idea] Bridge between eloquent and external API resources

### ISSUE THREAD

**Issue by georgeboot** [laravel/internals#823]
## TL;DR
Laravel Eloquent supports [API Resources](https://laravel.com/docs/5.5/eloquent-resources) since 5.5. This makes presenting data very easy for the API. The consumer of the API however, still needs to manually `Guzzle` the data in and make sense of it. These days, lot's of applications are built als a collection of micro services that communicate with each other.

Should't me make a Eloquent-like framework that has the power of Eloquent, but speaks to external (Laravel) API's rather then directly to a database?

## Background
I'm a huge fan on eloquent. The simplicity combined with the huge power makes it a tool I love to use. However, eloquent only works if you are storing your data directly to a database (MySQL, Postgres, etc.) In the area where micro services are becoming more popular by the day, I feel Laravel should make a move.

And Laravel has: since 5.5 eloquent supports the very handy [API Resources](https://laravel.com/docs/5.5/eloquent-resources). This eases up the part of transforming your data to fir your JSON scheme a lot.

The other side however - the consumer of the API - still has to manually GET and POST the data, process it, and present it inside the consumer application.

I recently converted a regular application into micro services. I found myself replacing all my Eloquent calls with basic API calls, forcing me to create a lot of extra complexity on the API side. This worked, but takes away all the awesome power Eloquent had before. In short, the current simple `json` data representation layer is not built for the complexity eloquent has to offer.

**I think Laravel can do better!**

## Hacking eloquent
I have messed around with creating a eloquent adapter that speaks to an API rather then to a database, but this does not work. Eloquent was not designed that way, and I doubt if fixing this would make any sense.

## The solution
I think Laravel should contain a framework for speaking to (Laravel) API resources with the power Eloquent has. Practically, this would involve the following:

1. Define a `json` scheme that supports all popular eloquent actions (searching, relations, etc).
2. Make a API resource type that takes care of this automatically. Just like the present `$fillable`, there should be a mechanism to limit actions.
3. Build a model type that uses the external datastore rather then a regular database. Perhaps the API should be the same as eloquent has now.
4. Make sure all this works just as dead simple as Laravel currently works and can be used without effort.

**Comment by Dylan-DPC** [comment]
I don't think this should be in the core. It will add excessive bloat as not every project wants to talk with an API.

**Comment by m1guelpf** [comment]
@georgeboot I had this idea soe time ago and asked @themsaid about his opinion, and he said this'd have to be in a package first to consider pulling it to core

**Comment by chrislandeza** [comment]
Agree with @Dylan-DPC. 

I'd love to see a good package for this one.

**Comment by m1guelpf** [comment]
@chrislandeza There's a little package I created that does something similar to this. It's currently on prerelease v0.1 and has no docs, but it may serve your propouse it you do a bit of source-diving. It's called [Laravel Whisper](https://github.com/laravel-whisper).

**Comment by georgeboot** [comment]
@Koohiisan not really. We currently use GraphQL and this tackles some of the difficulties, but not all.

We’ve also looked at gRPC but that has no server support for php (yet).

**Comment by mfn** [comment]
You already mentioned GraphQL but not which implementation, https://github.com/nuwave/lighthouse is one which _tightly_ integrates with Eloquent and they're working very had to improve things all the time.

**Comment by georgeboot** [comment]
@mfn yes, we indeed use Lighthouse
